quantum/backlight/backlight_arm.c

```diff
@@ -0,0 +1,261 @@
+    pwmStart(&BACKLIGHT_PWM_DRIVER, &pwmCFG_breathing);
+    chSysLockFromISR();
+    pwmEnablePeriodicNotification(&BACKLIGHT_PWM_DRIVER);
+    pwmEnableChannelI(
+      &BACKLIGHT_PWM_DRIVER,
+      0,
+      PWM_FRACTION_TO_WIDTH(
+        &BACKLIGHT_PWM_DRIVER,
+        0xFFFF,
+        0xFFFF
+      )
+    );
+    chSysUnlockFromISR();
+}
+
+void breathing_interrupt_disable(void){
+    pwmStop(&BACKLIGHT_PWM_DRIVER);
+    pwmStart(&BACKLIGHT_PWM_DRIVER, &pwmCFG);
+}
+
+void breathing_enable(void)
+{
+  breathing_counter = 0;
+  breathing_halt = BREATHING_NO_HALT;
+  breathing_interrupt_enable();
+}
+
+void breathing_pulse(void)
+{
+    if (get_backlight_level() == 0)
+      breathing_min();
+    else
+      breathing_max();
+    breathing_halt = BREATHING_HALT_ON;
+    breathing_interrupt_enable();
+}
+
+void breathing_disable(void)
+{
+    //printf("breathing_disable()\n");
+    breathing_interrupt_disable();
+    // Restore backlight level
+    backlight_set(get_backlight_level());
+}
+
+void breathing_self_disable(void)
+{
+  if (get_backlight_level() == 0)
+    breathing_halt = BREATHING_HALT_OFF;
+  else
+    breathing_halt = BREATHING_HALT_ON;
+}
+
+void breathing_toggle(void) {
+  if (is_breathing())
+    breathing_disable();
+  else
+    breathing_enable();
+}
+
+void breathing_period_set(uint8_t value)
+{
+  if (!value)
+    value = 1;
+  breathing_period = value;
+}
+
+void breathing_period_default(void) {
+  breathing_period_set(BREATHING_PERIOD);
+}
+
+void breathing_period_inc(void)
+{
+  breathing_period_set(breathing_period+1);
+}
+
+void breathing_period_dec(void)
+{
+  breathing_period_set(breathing_period-1);
+}
+
+/* To generate breathing curve in python:
+ * from math import sin, pi; [int(sin(x/128.0*pi)**4*255) for x in range(128)]
+ */
+static const uint8_t breathing_table[BREATHING_STEPS] = {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 2, 3, 4, 5, 6, 8, 10, 12, 15, 17, 20, 24, 28, 32, 36, 41, 46, 51, 57, 63, 70, 76, 83, 91, 98, 106, 113, 121, 129, 138, 146, 154, 162, 170, 178, 185, 193, 200, 207, 213, 220, 225, 231, 235, 240, 244, 247, 250, 252, 253, 254, 255, 254, 253, 252, 250, 247, 244, 240, 235, 231, 225, 220, 213, 207, 200, 193, 185, 178, 170, 162, 154, 146, 138, 129, 121, 113, 106, 98, 91, 83, 76, 70, 63, 57, 51, 46, 41, 36, 32, 28, 24, 20, 17, 15, 12, 10, 8, 6, 5, 4, 3, 2, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0};
+
+// Use this before the cie_lightness function.
+static inline uint16_t scale_backlight(uint16_t v) {
+  return v / BACKLIGHT_LEVELS * get_backlight_level();
+}
+
+static void breathing_callback(PWMDriver *pwmp)
+{
+  (void)pwmp;
+  uint16_t interval = (uint16_t) breathing_period * 256 / BREATHING_STEPS;
+  // resetting after one period to prevent ugly reset at overflow.
+  breathing_counter = (breathing_counter + 1) % (breathing_period * 256);
+  uint8_t index = breathing_counter / interval % BREATHING_STEPS;
+
+  if (((breathing_halt == BREATHING_HALT_ON) && (index == BREATHING_STEPS / 2)) ||
```

Comment:
this line is confusing. can't quite figure out what it does? any insight?

Reply:
ok, I have a theory. breathing is defined as 256 steps. Only 128 are defined for curve. There when the curve reaches end, it turns off breathing for 128 steps. unfortunately it looks like it keeps doing the calculation though

Reply:
looks like it's only used for pulse

Reply:
Optimising the current solution will be tackled as part of another round of "Improve ARM backlight support" where I hope to unify some of the breathing logic between ARM and AVR.